Song select filtering

ddevault opened the pull request:
Includes basic sorting+grouping UI, implementation to come separately.

ddevault committed:
Initial layout of song select filtering UI

ddevault committed:
Move filter container z depth and add hover fx

ddevault committed:
Rename FilterSongSelect, use OsuColour

ddevault committed:
Use default background on empty song select screen

ddevault committed:
Improve beatmap import

ddevault committed:
Use OsuColour for green text

ddevault committed:
Implement filtering with strings

ddevault committed:
Merge remote-tracking branch 'upstream/master' into song-select-filtering

ddevault committed:
Fixes following recent updates upstream

ddevault committed:
Merge remote-tracking branch 'upstream/master' into song-select-filtering

peppy commented on osu.Desktop/Program.cs:
```diff
@@ -25,17 +26,22 @@ public static class Program
+                        if (!importer.Import(Path.GetFullPath(file)).Wait(3000))
                             throw new TimeoutException(@"IPC took too long to send");
-                    Console.WriteLine(@"Sent import requests to running instance");
+                        Thread.Sleep(500);
```
why 500ms?

SirCmpwn replied:
Arbitrary number to slow it down, since the server doesn't tell us when it's done processing an import. We could refactor this but should be done separately imo.

peppy replied:
i really don't like arbitrary sleeps. can we just leave this out for now?

SirCmpwn replied:
Adding multiple beatmaps at once doesn't work if we do, which makes it pretty annoying to wipe and refill your database.

peppy replied:
i've added 100 at a time with no issues until now?

SirCmpwn replied:
Alright, w/e. I'll just keep it on my own builds.

peppy commented on osu.Desktop/Program.cs:
```diff
@@ -25,17 +26,22 @@ public static class Program
                 {
                     var importer = new BeatmapImporter(host);
-
+                    Directory.SetCurrentDirectory(cwd);
```
what's this for?

SirCmpwn replied:
DesktopGameHost changes the cwd to the path of the binary. This changes it back so relative paths make sense when passed at the command line.

peppy commented on osu.Desktop/Program.cs:
```diff
@@ -25,17 +26,22 @@ public static class Program
         {
             LegacyFilesystemReader.Register();
+            
+            var cwd = Directory.GetCurrentDirectory();
```
`Environment.CurrentDirectory` may be a better fit? not sure what you're using this for though. needs comments since it's not obvious.

SirCmpwn replied:
Will add comments.

peppy commented on osu.Game/Beatmaps/Drawables/BeatmapGroup.cs:
```diff
@@ -1,42 +1,46 @@
-            set
-            {
-                state = value;
+        public bool Hidden;
```
not sure i like this being a public bool with no actual logic behind it. changing this value should either not be allowed or should actually perform an action. consider making it private if it's just storing a local state.

SirCmpwn replied:
It tells the CarouselContainer how to treat it, primarily.

peppy commented on osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs:
```diff
@@ -56,6 +55,8 @@ public WorkingBeatmap Beatmap
         {
             Beatmap = beatmap;
+            if (beatmap == null)
+                Add(background = new Background(@"Backgrounds/bg1"));
```
move this logic to line29 and combine with the rest of the logic that is adding the background (and apply the same blur logic etc. it should behave exactly the same.

peppy commented on osu.Game/Screens/Select/CarouselContainer.cs:
```diff
@@ -248,6 +251,13 @@ protected override void Update()
             }
         }
 
+        public void InvalidateVisible()
```
what does this even mean? nothing is being invalidated here.

SirCmpwn replied:
Not invalidate in the sense of Drawable.Invalidate, but it says "the Y positions of panels is now wrong, the Hidden state of some of them has changed"

peppy commented on osu.Game/Screens/Select/FilterControl.cs:
```diff
@@ -0,0 +1,266 @@
+{
+    public class FilterControl : Container
+    {
+        public enum SortMode
```
i prefer enums going at the end of the file

peppy commented on osu.Game/Screens/Select/FilterControl.cs:
```diff
@@ -0,0 +1,266 @@
+    
+        public Action FilterChanged;
+
+        public string Search { get; private set; } = string.Empty;
```
any reason this needs storage as opposed to just linking directly to searchTextBox?

ie. 
`public string Search => searchTextBox.Text`

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -174,9 +187,55 @@ private void start()
+
+        private void filterChanged()
+        {
+            if (filterTask != null)
```
`filterTask?.Cancel()`

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -174,9 +187,55 @@ private void start()
+                foreach (var beatmapGroup in carousel)
+                {
+                    var set = beatmapGroup.BeatmapSet;
+                    if (set == null)
```
is this ever a thing?

SirCmpwn replied:
Just coding defensively. Will remove it.

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -174,9 +187,55 @@ private void start()
+                        || (set.Metadata.TitleUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1;
+                    if (match)
+                    {
+                        changed = changed && !beatmapGroup.Hidden;
```
`change &= !beatmapGroup.Hidden`

SirCmpwn replied:
&&= doesn't exist in C#. Does &= do the same thing?

Tom94 replied:
Yes. IIRC in C# && is literally &, only with short-circuiting.

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -174,9 +187,55 @@ private void start()
+                    {
+                        changed = changed && !beatmapGroup.Hidden;
+                        beatmapGroup.Hidden = false;
+                        beatmapGroup.Header.Alpha = 1;
```
ugly. make Hidden a property and actually do this inside the beatmapGroup.

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -174,9 +187,55 @@ private void start()
+                    {
+                        changed = changed && beatmapGroup.Hidden;
+                        beatmapGroup.Hidden = true;
+                        beatmapGroup.Header.Alpha = 0;
```
ditto

peppy commented:
- Escape key doesn't work at song select any more.
- Consider making the Hidden thing a state of `BeatmapGroupState` rather than this `bool Hidden` poop.

SirCmpwn commented:
>Consider making the Hidden thing a state of BeatmapGroupState rather than this bool Hidden poop.

I did consider that, but it complicates things, imo a bit too much.

ddevault committed:
Merge remote-tracking branch 'upstream/master' into song-select-filtering

ddevault committed:
Add comments around cwd handling

ddevault committed:
Update null background handling

ddevault committed:
Make BeatmapGroup.Hidden more intelligent

ddevault committed:
Refactor FilterControl.Search

ddevault committed:
Use null coalesce operator where appropriate

ddevault committed:
&&=, unnecessary null check

ddevault committed:
Relocate enums

ddevault committed:
Update osu-framework

ddevault committed:
SearchTextBox: Don't handle Key.{Enter,Left,Right}

ddevault committed:
Fix focus interactions with search text box

ddevault committed:
Only grab focus if nothing else has it

ddevault committed:
Refactor out Hidden

Also fixes a bug I didn't think about before

peppy commented on osu.Game/Screens/Select/SearchTextBox.cs:
```diff
@@ -0,0 +1,89 @@
+        {
+            if (GrabFocus && !HasFocus && IsVisible)
+            {
+                var inputManager = Parent;
```
i wouldn't bother adding this parent check for now. when it becomes an issue we should fix it in InputManager's Focused handling, not locally here.

ddevault committed:
Remove code to check for other focused controls

peppy commented:
- escape still doesn't work
- selecting any difficulty that isn't the first in a group then entering play mode causes weird shit to happen

please test stuff yourself :(

SirCmpwn commented:
>escape still doesn't work

Sorry. I did test this, but not after the last set of changes.

>selecting any difficulty that isn't the first in a group then entering play mode causes weird shit to happen

Cannot reproduce.

ddevault committed:
Merge remote-tracking branch 'upstream/master' into song-select-filtering

ddevault committed:
Fix escape key to exit PlaySongSelect

This is less than ideal but is the least disruptive solution. The
InputManager itself holds Escape keypresses from getting to anything
else if something is focused.

ddevault committed:
Fix issue with background on player

Not sure why this only happens on my branch, I didn't touch this code.

ddevault committed:
Fix extra click sound effect

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -189,9 +189,13 @@ private void start()
 
         private void filterChanged()
         {
+            if (previousFilter == filter.Search)
```
did you mean to use this?

![image](https://cloud.githubusercontent.com/assets/191335/22490860/04d5a970-e863-11e6-9b89-97b315f09347.png)

SirCmpwn replied:
No, but I forgot to do the other thing I meant to do that justified that decision.

ddevault committed:
Initialize previousFilter to string.Empty

ddevault committed:
Use OnCommit instead of OnChange simplifies things

peppy commented on osu.Game/Beatmaps/Drawables/BeatmapGroup.cs:
```diff
@@ -28,18 +28,21 @@ class BeatmapGroup : IStateful<BeatmapGroupState>
 
         private BeatmapGroupState state;
-
+        
```
trim

ddevault committed:
Minor cleanups

peppy commented on osu.Game/Beatmaps/Drawables/BeatmapGroup.cs:
```diff
@@ -48,14 +51,20 @@ public BeatmapGroupState State
                         break;
                     case BeatmapGroupState.Collapsed:
+                    case BeatmapGroupState.Hidden:
+                        if (state == BeatmapGroupState.Hidden && state != value)
```
`Header.Alpha = state == BeatmapGroupState.Hidden ? 0 : 1;`
put this outside the switch and remove all the conditionals you added inside it.

ddevault committed:
Fix poorly written BeatmapGroup logic

peppy committed:
Merge remote-tracking branch 'upstream/master' into song-select-filtering

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -167,9 +167,6 @@ protected override void OnEntering(GameMode last)
         {
             base.OnEntering(last);
 
-            (Background as BackgroundModeBeatmap)?.BlurTo(Vector2.Zero, 1000);
```
why is this removed? the background no longer unblurs

peppy commented on osu.Game/Screens/Select/CarouselContainer.cs:
```diff
@@ -133,10 +134,10 @@ private float computeYPositions()
 
                         //on first display we want to begin hidden under our group's header.
-                        if (panel.Alpha == 0)
+                        if (panel.Alpha == 0 && group.State != BeatmapGroupState.Hidden)
```
guaranteed to be true

SirCmpwn replied:
No, during animations Alpha could be 0.

peppy replied:
read this again

peppy commented on osu.Game/Screens/Select/CarouselContainer.cs:
```diff
@@ -133,10 +134,10 @@ private float computeYPositions()
                             panel.MoveToY(headerY);
 
-                        movePanel(panel, true, ref currentY);
+                        movePanel(panel, group.State != BeatmapGroupState.Hidden, ref currentY);
```
ditto

peppy commented on osu.Game/Screens/Select/CarouselContainer.cs:
```diff
@@ -300,5 +310,15 @@ public void SelectRandom()
             SelectGroup(group, panel);
         }
+
+        public IEnumerator<BeatmapGroup> GetEnumerator()
```
public IEnumerator<BeatmapGroup> GetEnumerator() => groups.GetEnumerator();

peppy commented on osu.Game/Screens/Select/CarouselContainer.cs:
```diff
@@ -300,5 +310,15 @@ public void SelectRandom()
+            return groups.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
```
IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

ddevault committed:
Use OnChange instead of OnCommit

OnCommit only gets fired when focus is lost

ddevault committed:
Tidy up GetEnumerable

ddevault committed:
Remove constant boolean expressions

ddevault committed:
Fix use of previous value in Alpha check

Frustratingly, this does not fix the problem of ghost groups

ddevault committed:
Replace incorrectly removed background logic

I had earlier thought that removing this would solve the weirdness with
starting the beatmap on any but the first difficulty, and forgot to
replace it when I found the actual cause.

peppy committed:
Fix laggy animations, incorrect hiding logic.

peppy committed:
Fix deselected group remaining visible after search.

peppy committed:
Fix keyboard controls not respecting search filter.

peppy committed:
Framework update.

peppy commented:
Posted fixes for remaining issues at https://github.com/SirCmpwn/osu/pull/1.

ddevault committed:
Merge pull request #1 from peppy/song-select-filtering-fixes

Fixes inconsistencies and shortcomings of filtering.

Tom94 commented on osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs:
```diff
@@ -22,17 +22,19 @@ public WorkingBeatmap Beatmap
                 {
-                    Background newBackground = new BeatmapBackground(beatmap);
+                    Background newBackground;
+                    if (beatmap == null)
```
I am against allowing beatmap == null. We should avoid the whole class of problems assiciated with null by having a default built-in beatmap. Less corner cases (if statements for things like null) means cleaner and more readable code.

SirCmpwn replied:
:-1: I would find that very annoying. A whole class of problems is really just 2 or 3 problems, which have all been addressed as of this PR.

Tom94 replied:
What I mean with "whole class of problems" is that _every single time a beatmap is used_ you need a special treatment for the null case. Every. Single. Time. In the old code base those were literally hundreds of times.

What do you think is more annoying than that about having a default beatmap that avoids all this special treatment of a null beatmap?

SirCmpwn replied:
But only on this screen. You can't get to the player if there's no song for you to select here. I'm not suggesting we allow every screen to have a null beatmap, it makes sense of course for Player to require a non-null beatmap for example.

SirCmpwn replied:
Unless you want to download, process, and import syncronously (blocking the UI for several seconds average case), there will be some time when the UI is interactive and the beatmap database is empty, and thus the user can find their way over to screens where we need to handle that.

Tom94 replied:
You can still prohibit things or add special behavior by checking `beatmap == Default`. In the majority of cases you don't need that, though. The beatmap could have the osu! theme as song, and would be built-in into the executable anyway, so you also avoid the logic needed to play the theme song when no beatmap is selected (which can become surprisingly annoying when all transitioning to beatmap-based audio need to be seamless).

SirCmpwn replied:
Checking if `beatmap == Default` is the same thing as `beatmap == null`, only sillier.

The beatmap still needs to be imported into the database and written into the filesystem, which takes time. If you don't want to write it in the filesystem, the maintenance cost and complexity overhead of having the beatmap store support some kind of virtual beatmap that's not sourced from the filesystem is nuts compared to just checking if there are no beatmaps in the db.

ddevault committed:
@Tom94 pointed out this problem

ddevault committed:
Merge branch 'master' into song-select-filtering

Tom94 commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -174,9 +188,43 @@ private void start()
+                        || (set.Metadata.ArtistUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.Title ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.TitleUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1;
+                    if (match)
```
The filter will also have to work with individual difficulties; not just groups; in the future. Probably out of score for this PR, too, given that we still need to finalize how sorting e.g. by difficulty will work in the first place.

peppy replied:
and be regex and support a lot more things, yep :p
